File: src/leadforge/llm/entity_resolution.py
import json

import structlog

from leadforge.llm.client import VLLMClient

logger = structlog.get_logger()

ENTITY_RESOLUTION_PROMPT = """Compare these two business records and determine if they are the same business.

Business A:
- Name: {name_a}
- Address: {address_a}
- Zip: {zip_a}
- Phone: {phone_a}

Business B:
- Name: {name_b}
- Address: {address_b}
- Zip: {zip_b}
- Phone: {phone_b}

Respond with ONLY a JSON object:
{{"is_match": true/false, "confidence": 0.0-1.0, "reason": "brief explanation"}}
"""

MERGE_THRESHOLD = 0.8


async def resolve_entities(
    record_a: dict, record_b: dict, client: VLLMClient | None = None
) -> dict:
    """Compare two business records using LLM and return match result.

    Returns: {"is_match": bool, "confidence": float, "reason": str}
    """
    prompt = ENTITY_RESOLUTION_PROMPT.format(
        name_a=record_a.get("name", ""),
        address_a=record_a.get("address", ""),
        zip_a=record_a.get("zip_code", ""),
        phone_a=record_a.get("phone", ""),
        name_b=record_b.get("name", ""),
        address_b=record_b.get("address", ""),
        zip_b=record_b.get("zip_code", ""),
        phone_b=record_b.get("phone", ""),
    )

    own_client = client is None
    if own_client:
        client = VLLMClient()

    try:
        response = await client.complete(prompt, max_tokens=200)
        if not response:
            return {"is_match": False, "confidence": 0.0, "reason": "LLM unavailable"}

        # Parse JSON response
        result = json.loads(response.strip())
        return {
            "is_match": result.get("is_match", False)
            and result.get("confidence", 0) >= MERGE_THRESHOLD,
            "confidence": result.get("confidence", 0.0),
            "reason": result.get("reason", ""),
        }
    except (json.JSONDecodeError, KeyError) as e:
        logger.warning("entity_resolution_parse_failed", error=str(e))
        return {"is_match": False, "confidence": 0.0, "reason": f"Parse error: {e}"}
    finally:
        if own_client:
            await client.close()


def merge_records(primary: dict, secondary: dict) -> dict:
    """Merge two business records, preferring non-null values from primary."""
    merged = dict(primary)
    for key, value in secondary.items():
        if key not in merged or merged[key] is None:
            merged[key] = value
    return merged

File: api/src/lib/llm/entity-resolution.ts
// =py llm/entity_resolution
import { stripFences, type LlmClient } from './client';

export interface BusinessRecord { name?: string | null; address?: string | null; zip_code?: string | null; phone?: string | null }

export interface EntityMatch { is_match: boolean; confidence: number; reason: string }

export const MERGE_THRESHOLD = 0.8;

const ENTITY_RESOLUTION_PROMPT = (a: BusinessRecord, b: BusinessRecord) => `Compare these two business records and determine if they are the same business.

Business A:
- Name: ${a.name ?? ''}
- Address: ${a.address ?? ''}
- Zip: ${a.zip_code ?? ''}
- Phone: ${a.phone ?? ''}

Business B:
- Name: ${b.name ?? ''}
- Address: ${b.address ?? ''}
- Zip: ${b.zip_code ?? ''}
- Phone: ${b.phone ?? ''}

Respond with ONLY a JSON object:
{"is_match": true/false, "confidence": 0.0-1.0, "reason": "brief explanation"}
`;

// =py resolve_entities: a match only counts when the model says so AND confidence >= MERGE_THRESHOLD
export async function resolveEntities(recordA: BusinessRecord, recordB: BusinessRecord, client: LlmClient): Promise<EntityMatch> {
  try {
    const response = await client.complete(ENTITY_RESOLUTION_PROMPT(recordA, recordB), { maxTokens: 200 });
    if (!response) return { is_match: false, confidence: 0, reason: 'LLM unavailable' };
    const result = JSON.parse(stripFences(response)) as Partial<EntityMatch>;
    // Python defaults a missing confidence to 0 but raises (and reports a parse error) on a non-numeric one.
    const confidence = 'confidence' in result ? result.confidence : 0;
    if (typeof confidence !== 'number' || !Number.isFinite(confidence)) throw new TypeError('confidence is not a number');
    return {
      is_match: Boolean(result.is_match) && confidence >= MERGE_THRESHOLD,
      confidence,
      reason: result.reason ?? '',
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.warn('entity_resolution_parse_failed', message);
    return { is_match: false, confidence: 0, reason: `Parse error: ${message}` };
  }
}

// =py merge_records: primary wins; secondary fills keys that are missing or null on primary
export function mergeRecords(primary: Record<string, unknown>, secondary: Record<string, unknown>): Record<string, unknown> {
  const merged: Record<string, unknown> = { ...primary };
  for (const [key, value] of Object.entries(secondary)) {
    if (!(key in merged) || merged[key] == null) merged[key] = value;
  }
  return merged;
}
